README.rst
ADAT Lightpipe Digital Audio Interface
......................................

:Latest release: 1.0.0alpha1
:Maintainer: henkmuller
:Description: ADAT Lightpipe Receiver and Transmitter


Key Features
============

* 48000 and 44100 ADAT receivers
* 48000 and 44100 ADAT transmitters
* Application for loopback testing on Simulator or HW

To Do
=====

* This software relies on the reference clock being 100 MHz, there is no out of the box version for non
100 Mhz reference clocks.

* ADAT Tx for 256x master clock (i.e. 48kHz from 12.288MHz master clock) not yet implemented  

Firmware Overview
=================

The modules in this repo implement an ADAT transmitter and receiver in a
core each. Separate cores are required to collect and supply data.

Known Issues
============

* None

Support
=======

At the discretion of the maintainer.

Required software (dependencies)
================================

  * None



CHANGELOG.rst
sc_adat Change Log
===================

1.0.0
-----

    * Initial release


searchindex.js
Search.setIndex({objects:{"":{adatReceiver48000:[0,0,1,""],adat_tx_port:[1,0,1,""],adat_tx:[1,0,1,""],adatReceiver44100:[0,0,1,""]}},terms:{set_port_clock:1,protocol:[1,3],edg:1,just:3,show:1,adatreceiver48000:0,thi:[0,1,3],over:[1,3],fall:1,cloverleaf:[0,1,2,3],eight:[0,1,3],known:0,code:[2,1,3],latter:1,adatreceiver44100:0,follow:[1,3],simultan:1,unbuff:1,paramet:[0,1],xai2:1,onli:[0,1,3],xs1:0,depend:[0,3],singl:[0,1,3],adat_tx_port:1,intermedi:1,xs1_port_1p:[0,1],send:1,should:[0,1],xs1_port_1o:1,codec:3,board:1,sampler:3,input:[0,1,3],main:[0,1],sent:1,applic:[0,3],non:0,collectsampl:0,"return":[0,1],thei:0,smux:1,read:[0,1],"void":[0,1],simplest:1,arriv:0,"break":0,transmit:[2,1,3],cannot:0,coaxial:3,desk:3,word:[0,1],bit:[0,1,3],characterist:3,align:1,requir:[0,1,3],enabl:1,success:0,p_data:1,signal:[0,1,3],revers:1,fewer:0,setupclock:1,separ:1,whether:3,token:1,api:[0,1,2],mhz:[0,1,3],either:[0,1,3],compris:0,each:0,newpag:[0,1,2,3],output:[0,1,3],zzzzzz:0,where:0,pdif:3,adat_ref:0,"48khz":0,compil:[0,2],"int":[0,1],set:1,often:1,cabl:3,captur:1,frame:0,some:3,direct:[2,1],transmitt:[1,3],xs1_clkblk_2:1,rate:[0,1,3],sure:1,sampl:[0,1,3],connect:1,happen:1,port:[0,1,2,3],after:[0,1],transmiss:[1,3],infinit:1,shown:0,defin:[0,2],microphon:3,micro:[0,1,2,3],assum:[0,1,3],adat_rx:0,normal:[0,3],mux:0,flop:[1,3],delet:0,state:0,exampl:[0,1,2,3],larg:0,suitabl:1,channel:[0,1,3],between:1,call:0,"import":3,plusmn:[0,1,2,3],gener:1,red:[0,1,2,3],refer:[0,1],core:1,inspect:0,run:3,"0x0zzzzzz0":0,power:3,ipod:3,adat_port:1,"0x000000z1":0,how:1,them:1,from:[0,1],"byte":1,pll:3,sinc:3,valu:[0,1,3],both:[0,1,3],thread:[0,1,2,3],frequenc:[0,3],could:0,precis:[1,3],prior:1,amount:3,out:1,synchronis:[0,1],section:1,rame:0,implement:3,parallel:[0,1],onto:[0,1],produc:1,block:1,first:[0,1],amber:[0,1,2,3],softwar:[2,3],simpli:[0,1],stream:[0,1,3],extra:[2,1],carri:[0,3],activ:1,modul:[0,1,3],number:[1,3],three:1,expect:0,mai:[0,3],instruct:1,alreadi:3,typic:3,complex:1,forget:1,merg:1,differ:1,masterclock:1,span:[0,1,2,3],usb:3,byterev:1,memori:3,chanend:[0,1],support:[0,3],system:3,"0xf":0,avail:3,start:1,program:[0,1],master:[1,3],includ:[0,3],handl:0,second:[0,1],interfac:[0,1],adat_tx:1,ochan:0,more:[1,3],"function":[0,1,2,3],nibbl:0,head:0,mck:1,option:0,termin:[0,1],handler:0,copi:1,xs1_ct_end:1,set_clock_fall_delai:1,provid:3,flag:1,part:3,pars:0,see:1,bull:[0,1,2,3],start_clock:1,c_port:1,line:0,exactli:0,than:3,present:0,must:1,"case":0,also:1,ident:[1,3],mck_blk:1,resynchronis:3,setup:1,flip:[1,3],par:[0,1],module_adat_tx:[2,3],crystal:[0,3],below:[0,1],can:[1,3],"9th":0,histor:0,driveport:1,otherwis:1,latch:1,mbox:[0,1,2,3],loop:0,result:0,control:1,fork:[0,1],khz:[0,3],c_data:1,mclk:1,pin:1,process:1,lock:0,chan:[0,1],digit:3,indic:0,itself:0,right:1,ihjsqueezeihj:[0,1,2,3],deliv:1,need:[0,1],everi:[0,1],outct:1,shift:0,unsign:0,avb:3,ratio:1,delai:1,recov:3,differenti:3,end:[0,1],inform:0,receiv:[0,2,3],anoth:1,make:1,format:1,when:0,same:[1,3],"while":[0,1],note:[0,1,3],mix:3,"100mhz":0,priori:0,speaker:3,take:1,which:[0,1],test:3,mip:3,document:2,receiveadat:0,noth:1,syncrhonis:1,higher:3,generatedata:1,begin:0,"1khz":0,resourc:3,user:0,optic:3,shall:1,multipli:1,sequenc:0,"class":[0,1,2,3],decod:0,statu:3,inuint:[0,1],buffer:[0,1],two:[0,1,3],extern:[1,3],multipl:3,oscil:3,data:[0,1,3],chang:0,outuint:1,other:[0,1,3],set_clock_src:1,appropri:1,dock:3,ani:[0,1,3],whole:0,clock:[0,1,3],adat:[0,1,2,3],drive:[1,3],doe:[0,3],issu:3,mark:0,declar:[0,1],green:[0,1,2,3],sect:[0,1,2,3],determin:3,time:[0,2],module_adat_rx:[2,3],audio:3,order:1,left:1},objtypes:{"0":"c:function"},titles:["ADAT Receive","ADAT Transmit","Documentation","ADAT software"],objnames:{"0":["c","function","C function"]},filenames:["api-rx","api-tx","index","summary"]})

module_adat_rx/src/adat_rx.h
// Copyright (c) 2011, XMOS Ltd, All rights reserved
// This software is freely distributable under a derivative of the
// University of Illinois/NCSA Open Source License posted in
// LICENSE.txt and at <http://github.xcore.com/>


/** ADAT Receive Thread (48kHz sample rate).
 *
 *  When a data rame is received, samples will be output onto the streaming channel
 *  At first a word 0x000000Z1 will be output, where ``Z`` are the user data; after
 *  that eight words 0x0ZZZZZZ0 will be output where ``ZZZZZZ`` is a 24-bit sample value.
 *  The eight words may refer to sample values on eight channels, or on fewer channels if
 *  muxing is used.
 *
 *  The function will return if it cannot lock onto a 48,000 Hz
 *  signal. Normally the 48000 function is called in a while(1) loop. If
 *  both 44,100 and 48,000 need to be supported, they should be called in
 *  sequence in a while(1) loop. Note that the functions are large, and
 *  that 44,100 should not be called if 44.1 KHz does not need to be supported.
 *
 *  \param p     ADAT port - should be 1-bit and clocked at 100MHz
 *  \param oChan channel on which decoded samples are output
 **/
void adatReceiver48000(buffered in port:32 p, chanend oChan);

/** ADAT Receive Thread (44.1kHz sample rate).
 *
 *  When a data rame is received, samples will be output onto the streaming channel
 *  At first a word 0x000000Z1 will be output, where ``Z`` are the user data; after
 *  that eight words 0x0ZZZZZZ0 will be output where ``ZZZZZZ`` is a 24-bit sample value.
 *  The eight words may refer to sample values on eight channels, or on fewer channels if
 *  muxing is used.
 *
 *  The function will return if it cannot lock onto a 44,100 Hz
 *  signal. Normally the 44,100 function is called in a while(1) loop. If
 *  both 44,100 and 48,000 need to be supported, they should be called in
 *  sequence in a while(1) loop. Note that the functions are large, and
 *  that 48,000 should not be called if 48 Khz does not need to be supported.
 *
 *  \param p     ADAT port - should be 1-bit and clocked at 100MHz
 *  \param oChan channel on which decoded samples are output
 **/

void adatReceiver44100(buffered in port:32 p, chanend oChan);


module_adat_tx/src/adat_tx.h
// Copyright (c) 2011, XMOS Ltd, All rights reserved
// This software is freely distributable under a derivative of the
// University of Illinois/NCSA Open Source License posted in
// LICENSE.txt and at <http://github.xcore.com/>

/* usage:
        buffered out port:32 adat_port;
        in port mck;
        clock mck_blk;
        par {
           {
             outuint(c_data, 1024 or 512 or 256);  // master clock multiplier
             outuint(c_data, 0 or 2 or 4);  // SMUX flag
             while (!done) {
               for (int i = 0; i < 8; i++) {
                 outuint(c_data, x);   // left aligned data (only 24 bits will be used)
               }
             }
             outct(c_data, XS1_CT_END);
           }
           adat_tx(c_data, c_port);
           {
             set_clock_src(mck_blk, mck);
             set_port_clock(adat_port, mck_blk);
             set_clock_fall_delay(mck_blk, 7);  // XAI2 board
             start_clock(mck_blk);
             while (1) {
               adat_port <: byterev(inuint(c_port));
             }
           }
        }
  why byte reverse? mixing word and byte channel outputs and channels are big endian
*/

/**Function that takes data over a channel end, and that outputs this in
 * ADAT format onto a 1-bit port. The 1-bit port should be clocked by the
 * master-clock, and an external flop should be used to precisely align the
 * edge of the signal to the master-clock.
 *
 * Data should be send onto c_data using outuint only, the first two values
 * should be The multiplier and the smux values, after that output any
 * number of eight samples (24-bit, right aligned), and if the process is
 * to be terminated send it an control token 1.
 *
 * The data is output onto a channel, which a separate process should
 * output to a port. This process should byte-reverse every word read over
 * the channel, and then output the reversed word to a buffered 1-bit port.
 *
 * \param   c_data   Channel over which to send sample values to the transmitter
 *
 * \param   c_port   Channel on which to generate the ADAT stream
 */
void adat_tx(chanend c_data, chanend c_port);


/**Function that takes data over a channel end, and that outputs this in
 * ADAT format onto a 1-bit port. The 1-bit port should be clocked by the
 * master-clock, and an external flop should be used to precisely align the
 * edge of the signal to the master-clock.
 *
 * Data should be send onto c_data using outuint only, the first two values
 * should be The multiplier and the smux values, after that output any
 * number of eight samples (24-bit, right aligned), and if the process is
 * to be terminated send it an control token 1.
 *
 * \param   c_data   Channel over which to send sample values to the transmitter
 *
 * \param   p_data   1-bit port on which to generate the ADAT stream
 */
void adat_tx_port(chanend c_data, buffered out port:32 p_data);



__generator_rx/adat.c
// Copyright (c) 2011, XMOS Ltd, All rights reserved
// This software is freely distributable under a derivative of the
// University of Illinois/NCSA Open Source License posted in
// LICENSE.txt and at <http://github.xcore.com/>

#include <stdio.h>
#include <stdlib.h>
#include <math.h>

putOut(int a, int b, char c) {
    static char line0[97], line1[97], line2[97];
    static int x = 0;
    line0[x] = a + '0';
    line1[x] = b + '0';
    line2[x] = c;
    if (x == 95) {
        fprintf(stderr, "%s\n%s\n%s\n\n", line0, line1, line2);
        x = 0;
    } else {
        x++;
    }
}

double sqr(double x) {
    return x*x;
}
int main(int argc, char *argv[]) {
    int readBits = 0;
    int shift = 3;
    int bitsLeft = 0;
    double firstBitIsAt;
    double bitsPassed;
    int rounded;
    int getRidOfBit = 0;
    int bitsInCompressed =0;
    double freq = 48;
    int channel = 0;              // set to zero for port.
    int i;
    int total = 0;
    int verbose = 0;
    int last = 0;
    unsigned int mask = 0x80808080;
    int samplePoint = 0;
    double bitPoints[256], bitTime, theBitPoints[256];
    int nextBitToTake = 0;
    //double referenceFrequency = 999375.0;
    double referenceFrequency =1000000.0;

    firstBitIsAt = 15;
    for(i = 1; i < argc; i++) {
        if(strcmp(argv[i], "-t") == 0) {
            channel = 1; // test
        } else if(strcmp(argv[i], "-v") == 0) {
            verbose = 1;
        } else if(strcmp(argv[i], "-4") == 0) {
            freq = 44.1; // different frequency
            mask = 0x80402010;
            shift = 1;
            firstBitIsAt = 20;
        } else {
            fprintf(stderr, "Usage: %s [-t] [-4]\n", argv[0]);
            exit(1);
        }
    }
    bitTime =  referenceFrequency/(freq*256);
    bitPoints[0] = bitTime/2;
    for(i = 1; i < 256; i++) {
        bitPoints[i] = bitPoints[i-1] + bitTime;
    }
    printf("// GENERATED CODE - DO NOT EDIT\n");
    printf("// Comments are in the generator \n");
    printf("// Generated for devices with a reference clock of %10.6f Mhz\n", referenceFrequency/10000.0);
    printf("// If both 48000 and 44100 are to be supported, then\n// call adatReceiver48000 and 44100 in a while(1) loop\n");
    printf("#include <xs1.h>\n");
    printf("#include <stdio.h>\n");
    if (channel) {
        printf("extern void output(unsigned x, unsigned y, int z);\n");
    }
    printf("#pragma unsafe arrays\n");
    printf("void adatReceiver%d(%s p, chanend oChan) {\n", freq > 46 ? 48000 : 44100, channel ? "streaming chanend" : "buffered in port:32");
    printf("    const unsigned int mask = 0x%08x;\n", mask);
    printf("    unsigned compressed;\n");
    printf("    unsigned nibble, word = 1, fourBits, data;\n");
    printf("    int old, violation;\n");
    printf("    unsigned int lookupCrcF[16] = ");
    if (freq > 46 ) {
        printf("{8, 9, 12, 13, 7, 6, 3, 2, 10, 11, 14, 15, 5, 4, 1, 0};\n");
    } else {
        printf("{8, 12, 10, 14, 9, 13, 11, 15, 7, 3, 5, 1, 6, 2, 4, 0};\n");
    }
    printf("    unsigned int lookupNRTZ[32] = {0, 8, 12, 4, 6, 14, 10, 2, 3, 11, 15, 7, 5, 13, 9, 1,\n");
    printf("                          1, 9, 13, 5, 7, 15, 11, 3, 2, 10, 14, 6, 4, 12, 8, 0};\n");
    printf("    for(int i =  0; i < 32; i++) { lookupNRTZ[i] <<= 4; }\n");
// This section of code will wait for a preamble - at least two words
// of all zeros or all ones.
    printf("    do {\n");
    printf("        old = word; p :> word;\n");
    printf("    } while (word != old || (word != 0 && word+1 != 0));\n");
    printf("    while(1) {\n");
// Invariant: the variable word contains the current line state - in the violation.
    printf("        violation = word;\n");

// Skip the rest of the violation - this will consume a single bit on the edge of the first sample.
    if (channel) {
        printf("        while (word == violation) { p :> word;}\n");
    } else {
        printf("        p when pinsneq(violation) :> int _;\n");
        printf("        p :> word;\n");
    }
    while (readBits < 8*24+4) {
// Loop invarariant: "word" contains a sample with 32 bits.
// Figure out optimal shift

        fprintf(stderr, "Next bit: %d expected at %f, we are at %f\n", nextBitToTake, bitPoints[nextBitToTake], firstBitIsAt);
        double maxError = 9999;
        unsigned theMask;
        for(i = -7; i < 8; i++) {
            int m = i > 0 ? mask >> i : mask << -i;
            int sampleNum = nextBitToTake;
            double error = 0;
            int total = 0;
            int j;
            for(j = 0; j < 32; j++) {
                if (m >> j & 1) {
                    double thisSample = firstBitIsAt +  j * 10 ;
                    error += sqr(bitPoints[sampleNum] - thisSample);
                  //  if (verbose) fprintf(stderr, "Mask %08x shift %d bit %d error %f\n", mask, i, j, error);
                    sampleNum++;
                    total++;
                }
            }
            error = sqrt(error/total);
          //  if (verbose) fprintf(stderr, "Mask %08x shift %d total error %f\n", mask, i, error);
            if (error < maxError) {
                maxError = error;
                shift = i;
                bitsInCompressed = total;
                theMask = m;
            }
        }
        if (shift < 0) {
            shift += 8;
        }
        if (verbose) {
            fprintf(stderr, "Shift %d for best error of %f (%d bits)\n", shift, maxError, bitsInCompressed);
            int sampleNum = nextBitToTake;
            int j;
            for(j = 0; j < 32; j++) {
                if (theMask >> j & 1) {
                    double thisSample = firstBitIsAt +  j * 10 ;
                    theBitPoints[sampleNum] = thisSample;
                    sampleNum++;
                }
            }

        }
        nextBitToTake += bitsInCompressed;

// The crc instruction compresses the four masked bits into a 4 bit number, the lookup
// table reconstructs the original four bits.
        printf("        fourBits = (word << %d) & mask;\n", shift);
   //     printf("        output(word, mask>> %d, %d);\n", shift, bitsInCompressed);
   //     printf("        printf(\"Extracted bits %%08x\\n\", fourBits);\n");
        printf("        crc32(fourBits, 0xf, 0xf);\n");
   //     printf("        printf(\"CRC-ed bits %%02x\\n\", fourBits);\n");
        printf("        compressed = lookupCrcF[fourBits];\n");
        if (bitsInCompressed < 4) {
// This code gets rid of a duplicated bit - whenever the mask shifts over the edge of the word,
// one bit will be duplicated. This means that we should drop one bit (the oldest), and use only
// three bits on this occasion
            printf("        compressed = compressed >> 1;\n");
        }
   //     printf("        printf(\"Looked up bits %%02x (%d bits)\\n\", compressed);\n", bitsInCompressed);
        if (verbose) {
            for(i = 0; i < 32; i++) {
                if ((mask >> shift) >> i & 1) {
                    if (bitsInCompressed == 4 || i > 7) {
                        int bitnum = total + i;
                        fprintf(stderr, "Bit %d sample %d (%d)%s\n", samplePoint, bitnum, bitnum-last, samplePoint % 5 == 0 ? " (Stuff bit)" : "");
                        last = bitnum;
                        samplePoint++;
                    }
                }
            }
            total += 32;
        }
        if (bitsLeft + bitsInCompressed > 4) {
// There are sufficient bits to form a nibble. The lookuptable decodes the NRTZ encoding.
// THe shift and or operations take the past few bits and the current bits and glue them together.
// The rest of the bits (that were masked off by the &31) are kept in the bottom bits of old.
  //          printf("        printf(\">>> Last 8 bits bits %%02x\\n\", (old | (compressed << %d)));\n", bitsLeft);
            printf("        nibble = lookupNRTZ[(old | (compressed << %d)) & 31];\n", bitsLeft);
   //         printf("        printf(\">>> Nibbled bits %%02x\\n\", nibble);\n");
            printf("        old = compressed >> %d;\n", (5-bitsLeft));
            readBits+=4;
            bitsLeft = bitsLeft - 5 + bitsInCompressed;
            if (readBits == 4) {
                printf("        outuint(oChan, nibble << 4 | 1);\n");
            } else if ((readBits-4) % 24 == 0) {
                printf("        data = (data | nibble) << 4;\n");
                printf("        outuint(oChan, data);\n");
            } else if ((readBits-8) % 24 == 0) {
                printf("        data = nibble << 4;\n");
            } else {
                printf("        data = (data | nibble) << 4;\n");
            }
        } else if (bitsLeft == 0) {
            printf("        old = compressed;\n");
            bitsLeft+=bitsInCompressed;
        } else { // odd case - a bit left and 3 compressed bits, only 4, collect in old.
            printf("        old = old | compressed << %d;\n", bitsLeft);
            bitsLeft+=bitsInCompressed;
        }
        firstBitIsAt += 320;
        printf("        p :> word;\n");
    }
    printf("        if (word != 0 && word+1 != 0) return;\n");
    printf("    }\n");
    printf("}\n");
    printf("\n");
    if (verbose) {
        for(i = 0; i < 256; i++) {
            fprintf(stderr, "Bit %d  should be at %f but at %f\n", i, bitPoints[i], theBitPoints[i]);
        }
        int state0 = 0;
        int state1 = 0;
        int k = 0, k1 = 0, k2 = 0; char c;
        for(i = 0; i < bitPoints[255]; i+=10) {
            if (theBitPoints[k2] == i) {
                c = '^';
                k2++;
            } else {
                c = ' ';
            }
            putOut(state0, state1, c);
            if ((bitPoints[k]+bitPoints[k+1]-10)/2 <= i) {
                state0 = !state0;
                k++;
            }
            if ((bitPoints[k1]+bitPoints[k1+1]+10)/2 <= i) {
                state1 = !state1;
                k1++;
            }
        }
    }
    return 0;
}
